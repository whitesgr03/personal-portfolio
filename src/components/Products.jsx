import PropTypes from "prop-types";

import { forwardRef, useState, useLayoutEffect } from "react";

import { getImageSize } from "../utils/handleGetImageSize";

import openNewTabIcon from "../image/open-in-new-icon.png";
import githubIcon from "../image/github-icon.png";

const Products = forwardRef(
	({ onShowProduct, productsImage, showModal, onLoad, onLoading }, ref) => {
		const [imageSize, setImageSize] = useState(null);

		useLayoutEffect(() => {
			const IMAGES_SIZE = [
				{
					device: "small",
					width: 628,
					height: 419,
				},
				{
					device: "large",
					width: 867,
					height: 578,
				},
			];

			const getSize = target =>
				IMAGES_SIZE.find(item => item.device === target);

			const handleSetImageSize = () => {
				const viewPortWidth = Math.max(
					document.documentElement.clientWidth || 0,
					window.innerWidth || 0
				);
				setImageSize(
					viewPortWidth < 1024 ? getSize("small") : getSize("large")
				);
			};

			handleSetImageSize();

			window.addEventListener("resize", handleSetImageSize);

			return () => {
				window.removeEventListener("resize", handleSetImageSize);
			};
		}, []);

		return (
			<div id="products" className="products" ref={ref}>
				<h2>My Works</h2>
				<ul>
					{productsImage.map((product, index) => (
						<li key={product.id}>
							<button
								type="button"
								tabIndex={showModal ? "-1" : ""}
								aria-label="buttonForImage"
								onClick={() => onShowProduct(product)}
							>
								<div className="imageWrap">
									{index !== 0 && (
										<div className="imageLoading"></div>
									)}
									<img
										onLoad={e => {
											index !== 0
												? onLoad(e)
												: onLoading(false);
										}}
										sizes=" (min-width: 1024px) 30vw, (min-width: 900px) 60vw, (min-width: 768px) 70vw, (min-width: 600px) 80vw, 100vw"
										srcSet={`${product.image.preview.phone_small.url} 628w, ${product.image.preview.phone_large.url} 867w, ${product.image.preview.tablet.url} 1068w, ${product.image.preview.desktop_small.url} 1235w,  ${product.image.preview.laptop.url} 1378w,  ${product.image.preview.desktop_medium.url} 1561w,  ${product.image.preview.desktop_large.url} 1672w`}
										src={
											product.image.preview.phone_small
												.url
										}
										alt={product.name}
										loading={index > 0 ? "lazy" : "eager"}
										decoding={index > 0 ? "async" : "auto"}
										width={imageSize && imageSize.width}
										height={imageSize && imageSize.height}
									/>
								</div>
							</button>

							<article>
								<h3>
									<button
										type="button"
										tabIndex={showModal ? "-1" : ""}
										aria-label="buttonForTitle"
										onClick={() => onShowProduct(product)}
									>
										{product.name}
									</button>
								</h3>

								<p>
									Lorem ipsum dolor sit amet consectetur
									adipisicing elit. Perspiciatis reprehenderit
									adipisci tenetur, rerum impedit ex atque
									fugiat repudiandae ad ipsum temporibus
									totam! Repudiandae iure nisi quidem unde
									laudantium, totam nobis!
								</p>

								<div className="link">
									<a
										href="#live"
										target="_blank"
										tabIndex={showModal ? "-1" : ""}
									>
										Live <span>(opens in new tab)</span>
										<img
											src={openNewTabIcon}
											alt=""
											width="20"
											height="20"
										/>
									</a>
									<a
										href="#source"
										target="_blank"
										tabIndex={showModal ? "-1" : ""}
									>
										Source <span>(opens in new tab)</span>
										<img
											src={githubIcon}
											alt=""
											width="20"
											height="20"
										/>
									</a>
								</div>
							</article>
						</li>
					))}
				</ul>
			</div>
		);
	}
);

Products.propTypes = {
	onShowProduct: PropTypes.func,
	productsImage: PropTypes.array,
	showModal: PropTypes.bool,
	onLoad: PropTypes.func,
	onLoading: PropTypes.func,
};

export default Products;
