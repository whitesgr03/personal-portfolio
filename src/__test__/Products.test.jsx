import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

import { fireEvent } from "@testing-library/react";

import Products from "../components/Products";

describe("Renders Products Component", () => {
	it("Should return Modal DOM with product", () => {
		const mockProductsImage = [
			{
				id: "project1",
				name: "project1",
				image: {
					preview: {
						phone: {
							url: "...",
						},
						tablet: {
							url: "...",
						},
						laptop: {
							url: "...",
						},
						desktop: {
							url: "...",
						},
					},

					pictures: [
						{
							id: "project1-picture1",
							phone: {
								url: "...",
							},
							tablet: {
								url: "...",
							},
							laptop: {
								url: "...",
							},
							desktop: {
								url: "...",
							},
						},
						{
							id: "project1-picture2",
							phone: {
								url: "...",
							},
							tablet: {
								url: "...",
							},
							laptop: {
								url: "...",
							},
							desktop: {
								url: "...",
							},
						},
						{
							id: "project1-picture3",
							phone: {
								url: "...",
							},
							tablet: {
								url: "...",
							},
							laptop: {
								url: "...",
							},
							desktop: {
								url: "...",
							},
						},
					],
				},
			},
			{
				id: "project2",
				name: "project2",
				image: {
					preview: {
						phone: {
							url: "...",
						},
						tablet: {
							url: "...",
						},
						laptop: {
							url: "...",
						},
						desktop: {
							url: "...",
						},
					},
					pictures: [
						{
							id: "project2-picture1",
							phone: {
								url: "...",
							},
							tablet: {
								url: "...",
							},
							laptop: {
								url: "...",
							},
							desktop: {
								url: "...",
							},
						},
						{
							id: "project2-picture2",
							phone: {
								url: "...",
							},
							tablet: {
								url: "...",
							},
							laptop: {
								url: "...",
							},
							desktop: {
								url: "...",
							},
						},
						{
							id: "project2-picture3",
							phone: {
								url: "...",
							},
							tablet: {
								url: "...",
							},
							laptop: {
								url: "...",
							},
							desktop: {
								url: "...",
							},
						},
					],
				},
			},
		];

		render(<Products productsImage={mockProductsImage} />);

		const [firstButtonForImage] = screen.getAllByRole("button", {
			name: "buttonForImage",
		});

		const [firstButtonForTitle] = screen.getAllByRole("button", {
			name: "buttonForTitle",
		});

		expect(firstButtonForImage).toBeInTheDocument();
		expect(firstButtonForTitle).toBeInTheDocument();
	});
	it("Should show modal with product on click", async () => {
		const user = userEvent.setup();
		const mockOnShowProduct = jest.fn();
		const mockProductsImage = [
			{
				id: "project1",
				name: "project1",
				image: {
					preview: {
						phone: {
							url: "...",
						},
						tablet: {
							url: "...",
						},
						laptop: {
							url: "...",
						},
						desktop: {
							url: "...",
						},
					},

					pictures: [
						{
							id: "project1-picture1",
							phone: {
								url: "...",
							},
							tablet: {
								url: "...",
							},
							laptop: {
								url: "...",
							},
							desktop: {
								url: "...",
							},
						},
						{
							id: "project1-picture2",
							phone: {
								url: "...",
							},
							tablet: {
								url: "...",
							},
							laptop: {
								url: "...",
							},
							desktop: {
								url: "...",
							},
						},
						{
							id: "project1-picture3",
							phone: {
								url: "...",
							},
							tablet: {
								url: "...",
							},
							laptop: {
								url: "...",
							},
							desktop: {
								url: "...",
							},
						},
					],
				},
			},
			{
				id: "project2",
				name: "project2",
				image: {
					preview: {
						phone: {
							url: "...",
						},
						tablet: {
							url: "...",
						},
						laptop: {
							url: "...",
						},
						desktop: {
							url: "...",
						},
					},
					pictures: [
						{
							id: "project2-picture1",
							phone: {
								url: "...",
							},
							tablet: {
								url: "...",
							},
							laptop: {
								url: "...",
							},
							desktop: {
								url: "...",
							},
						},
						{
							id: "project2-picture2",
							phone: {
								url: "...",
							},
							tablet: {
								url: "...",
							},
							laptop: {
								url: "...",
							},
							desktop: {
								url: "...",
							},
						},
						{
							id: "project2-picture3",
							phone: {
								url: "...",
							},
							tablet: {
								url: "...",
							},
							laptop: {
								url: "...",
							},
							desktop: {
								url: "...",
							},
						},
					],
				},
			},
		];

		render(
			<Products
				onShowProduct={mockOnShowProduct}
				productsImage={mockProductsImage}
			/>
		);

		const [firstButtonForImage] = screen.getAllByRole("button", {
			name: "buttonForImage",
		});

		await user.click(firstButtonForImage);

		expect(mockOnShowProduct).toBeCalledTimes(1);

		const [firstButtonForTitle] = screen.getAllByRole("button", {
			name: "buttonForTitle",
		});

		await user.click(firstButtonForTitle);

		expect(mockOnShowProduct).toBeCalledTimes(2);
	});
	it("Should press the Tab key to focus buttons and links", async () => {
			{
				id: "project1",
				name: "project1",
				image: {
					preview: {
						url: {
							phone: "https://fakeimg.pl/480x320/195c86/",
							table: "https://fakeimg.pl/1004x669/195c86/",
							laptop: "https://fakeimg.pl/1372x915/195c86/",
							desktop: "https://fakeimg.pl/1820x1213/195c86/",
						},
					},
					pictures: [
						{
							id: "project1-picture1",
							url: {
								phone: "https://fakeimg.pl/480x320/195c86/",
								table: "https://fakeimg.pl/1004x669/195c86/",
								laptop: "https://fakeimg.pl/1372x915/195c86/",
								desktop: "https://fakeimg.pl/1820x1213/195c86/",
							},
						},
						{
							id: "project1-picture2",
							url: {
								phone: "https://fakeimg.pl/480x320/195c86/",
								table: "https://fakeimg.pl/1004x669/195c86/",
								laptop: "https://fakeimg.pl/1372x915/195c86/",
								desktop: "https://fakeimg.pl/1820x1213/195c86/",
							},
						},
						{
							id: "project1-picture3",
							url: {
								phone: "https://fakeimg.pl/480x320/195c86/",
								table: "https://fakeimg.pl/1004x669/195c86/",
								laptop: "https://fakeimg.pl/1372x915/195c86/",
								desktop: "https://fakeimg.pl/1820x1213/195c86/",
							},
						},
					],
				},
			},
		];

		render(
			<Products allProducts={mockAllProducts} showModal={mokeShowModal} />
		);

		const [firstButtonForImage] = screen.getAllByRole("button", {
			name: "buttonForImage",
		});
		const [firstButtonForTitle] = screen.getAllByRole("button", {
			name: "buttonForTitle",
		});
		const [firstLinkForLive] = screen.getAllByRole("link", {
			name: "Live (opens in new tab)",
		});
		const [firstLinkForSource] = screen.getAllByRole("link", {
			name: "Source (opens in new tab)",
		});

		expect(firstButtonForImage).toHaveAttribute("tabIndex", "-1");
		expect(firstButtonForTitle).toHaveAttribute("tabIndex", "-1");
		expect(firstLinkForLive).toHaveAttribute("tabIndex", "-1");
		expect(firstLinkForSource).toHaveAttribute("tabIndex", "-1");
	});
});
