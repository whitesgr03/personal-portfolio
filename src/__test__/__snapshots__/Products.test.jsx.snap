// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders Products Component should return Products DOM 1`] = `
<div>
  <div
    class="products"
    id="products"
  >
    <h2>
      My 
      <span>
        Works
      </span>
    </h2>
    <ul>
      <li>
        <div
          class="wrap"
        >
          <button
            aria-label="buttonForImage"
            type="button"
          >
            <img
              alt="product1"
              sizes="(max-width: 1920px) 100vw, 1920px"
              src="https://fakeimg.pl/1920x1280/195c86/"
              srcset="https://fakeimg.pl/480x320/195c86/ 480w, https://fakeimg.pl/1004x669/195c86/ 1004w,https://fakeimg.pl/1372x915/195c86/ 1372w,https://fakeimg.pl/1820x1213/195c86/ 1820w"
            />
          </button>
        </div>
        <article>
          <h3>
            <button
              aria-label="buttonForTitle"
              type="button"
            >
              product1
            </button>
          </h3>
          <p>
            Lorem ipsum dolor sit amet consectetur adipisicing elit. Perspiciatis reprehenderit adipisci tenetur, rerum impedit ex atque fugiat repudiandae ad ipsum temporibus totam! Repudiandae iure nisi quidem unde laudantium, totam nobis!
          </p>
          <div
            class="link"
          >
            <a
              href="#live"
              target="_blank"
            >
              Live
              <img
                alt="live (opens in new tab)"
                src="open-in-new-icon.png"
              />
            </a>
            <a
              href="#source"
              target="_blank"
            >
              Source
              <img
                alt="github (opens in new tab)"
                src="github-icon.png"
              />
            </a>
          </div>
        </article>
      </li>
      <li>
        <div
          class="wrap"
        >
          <button
            aria-label="product2Img"
            type="button"
          >
            <img
              alt="product2"
              sizes="(max-width: 1920px) 100vw, 1920px"
              src="https://fakeimg.pl/1920x1280/195c86/"
              srcset="https://fakeimg.pl/480x320/195c86/ 480w, https://fakeimg.pl/1004x669/195c86/ 1004w,https://fakeimg.pl/1372x915/195c86/ 1372w,https://fakeimg.pl/1820x1213/195c86/ 1820w"
            />
          </button>
        </div>
        <article>
          <h3>
            <button
              type="button"
            >
              product2
            </button>
          </h3>
          <p>
            Lorem ipsum dolor, sit amet consectetur adipisicing elit. Quisquam optio quis dicta laborum reprehenderit iste asperiores doloremque labore molestiae delectus?
          </p>
          <div
            class="link"
          >
            <a
              href="#live"
              target="_blank"
            >
              Live
              <img
                alt="live (opens in new tab)"
                src="open-in-new-icon.png"
              />
            </a>
            <a
              href="#source"
              target="_blank"
            >
              Source
              <img
                alt="github (opens in new tab)"
                src="github-icon.png"
              />
            </a>
          </div>
        </article>
      </li>
    </ul>
  </div>
</div>
`;
